### PR TITLE
add maximum hpugraphs and disable_tensor_cache arguments to GaudiTrainer

### PR DESCRIPTION
**Pull request by skaulintel**
… trainer.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/transformers/training_args.py
Comment (libinta) [inline optimum/habana/transformers/training_args.py:163]: default=10

File: optimum/habana/transformers/training_args.py
Comment (libinta) [inline optimum/habana/transformers/training_args.py:99]: maximum graphs which will be cached. Reduce to save device memory

File: optimum/habana/transformers/training_args.py
Comment (libinta) [inline optimum/habana/transformers/training_args.py:97]: If True, tensors won't be cached in hpu graph and memory can be saved.

File: optimum/habana/transformers/trainer.py
Comment (libinta) [inline optimum/habana/transformers/trainer.py:1861]: max_graph=10

File: optimum/habana/transformers/trainer.py
Comment (libinta) [inline optimum/habana/transformers/trainer.py:1480]: max_graph=10

Comment (regisss) [inline optimum/habana/transformers/trainer.py:1480]:
```suggestion
                model = wrap_in_hpu_graph(model, disable_tensor_cache=args.disable_tensor_cache_hpu_graphs, max_graphs=args.max_hpu_graphs)
```
I guess we want to use the new training args you defined.

File: optimum/habana/transformers/trainer.py
Comment (regisss) [inline optimum/habana/transformers/trainer.py:1861]:
```suggestion
                model = wrap_in_hpu_graph(model, disable_tensor_cache=args.disable_tensor_cache_hpu_graphs, max_graphs=args.max_hpu_graphs)
```

File: optimum/habana/transformers/training_args.py
Comment (skaulintel) [reply optimum/habana/transformers/training_args.py:163]: default is None.

File: optimum/habana/transformers/trainer.py
Comment (regisss) [reply optimum/habana/transformers/trainer.py:1861]: Well, it's actually a comment from Transformers so that means that this part of the code will be removed in Transformers v5 (which we don't know when it will be released, there is no ETA). In the meantime, we still have to maintain it so let's keep the arguments here.

File: optimum/habana/transformers/training_args.py
Comment (regisss) [inline optimum/habana/transformers/training_args.py:155]:
```suggestion
    disable_tensor_cache_hpu_graphs: Optional[bool] = field(
```

**Review by regisss — COMMENTED** [review]
You will need to run the following to get the code style right:
```
pip install --upgrade black ruff
```

Is this PR still a WIP?

**Comment by skaulintel** [comment]
Removed the WIP label and verified on my end with the following command. Let me know if there is anything left I should add before merge.

`python run_qa.py   --model_name_or_path bert-large-uncased-whole-word-masking   --gaudi_config_name Habana/bert-large-uncased-whole-word-masking   --dataset_name squad   --do_train   --do_eval   --per_device_train_batch_size 24   --per_device_eval_batch_size 8   --learning_rate 3e-5   --num_train_epochs 2 --max_steps 40  --max_seq_length 384   --doc_stride 128   --output_dir /tmp/squad/   --use_habana   --use_lazy_mode   --use_hpu_graphs_for_inference   --throughput_warmup_steps 3 --disable_tensor_cache True --max_hpu_graphs 1 --overwrite_output_dir`

File: optimum/habana/transformers/training_args.py
Comment (regisss) [inline optimum/habana/transformers/training_args.py:99]:
```suggestion
            Maximum number of hpu graphs to be cached. Reduce to save device memory.
```

**Review by regisss — COMMENTED** [review]
Regarding unit tests, the best would be to add it to one of the regression test. Is there a task and a model where it makes sense to use it? Maybe BERT Large QA as in the command you provided?

**Comment by skaulintel** [comment]
@regisss Not sure if we want to change the existing bert large test to use max_graphs and disable_tensor_cache by default. Should we add a new test there? Alternatively, I see a test_hpu_graphs function here https://github.com/huggingface/optimum-habana/blob/e8b1b4e86960f342334f3fc020e5cb1ce54e1807/tests/test_trainer.py#L684

maybe we could add the new arguments here?

@libinta What do you suggest?

**Comment by regisss** [comment]
> @regisss Not sure if we want to change the existing bert large test to use max_graphs and disable_tensor_cache by default. Should we add a new test there? Alternatively, I see a test_hpu_graphs function here
> 
> https://github.com/huggingface/optimum-habana/blob/e8b1b4e86960f342334f3fc020e5cb1ce54e1807/tests/test_trainer.py#L684
> 
> maybe we could add the new arguments here?
> 
> @libinta What do you suggest?

Sounds good, let's add that to `test_hpu_graphs` :+1:

**Comment by regisss** [comment]
@skaulintel Why setting the default to 10 for `max_hpu_graphs`? Is it the same default in Synapse?

**Comment by skaulintel** [comment]
> @skaulintel Why setting the default to 10 for `max_hpu_graphs`? Is it the same default in Synapse?

Yes, it is.

**Review by regisss — APPROVED** [review]
LGTM!

The doc build failed but it's not related to this PR and it has been solved in main already. Let's wait for @libinta's approval and then we merge.

**Review by libinta — APPROVED** [review]

**Comment by regisss** [comment]
> > @skaulintel Why setting the default to 10 for `max_hpu_graphs`? Is it the same default in Synapse?
> 
> Yes, it is.

@skaulintel Actually it is 10 for the `ModuleCacher`, but it is `None` for `wrap_in_hpu_graph`. I'm going to change the default value to `None` then.

**Pull request merged**
